### ISSUE TITLE
seems to be some confusion with mapping files

### ISSUE THREAD

**Issue by ThosRTanner** [include-what-you-use/include-what-you-use#224]
I have this in my QT mapping file:

    { symbol: [ "QKeyEvent",      "private", "<QKeyEvent>", "public" ] },
    { symbol: [ "QMouseEvent", "private", "<QMouseEvent>", "public" ] },

in order to deal with the way QT defines these in the documentation as opposed to implementing them in the headers. However, this occasionally causes include-what-you-use to produce this output:

    x.cpp should add these lines:
    #include <QKeyEvent>          // for QKeyEvent
    #include <QMouseEvent>        // for QMouseEvent
    class QKeyEvent;
    class QMouseEvent;

which seems wrong.

**Comment by kimgr** [comment]
Yes, that does seem incorrect.

Can you reduce it to a free-standing translation unit, and post the IWYU output at verbose level 7? Thanks.

**Comment by ThosRTanner** [comment]
will try and get something minimal. Note for self: use -Xiwyu -verbosity= (not documented...)

**Comment by kimgr** [comment]
Thanks!

(I think it is, see e.g. `include-what-you-use --help` or [somewhat veiled] https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/InstructionsForDevelopers.md#debugging)

**Comment by ThosRTanner** [comment]
I'm adding the command line, source and .imp file now. The output file with --verbose=7 might take a bit longer

OH. No I'm not. I don't have write permission. Here they are in line. Or at least the short ones. given the huge amount of output generated any idea what I should do with that?

    #include <QDialog>

    class BrowserDialog : public QDialog
    {
      virtual bool eventFilter(QObject *object, QEvent *event);
    };

    #include <QKeyEvent>

    bool BrowserDialog::eventFilter(QObject *object, QEvent *event)
    {
      if (event->type() == QEvent::KeyPress) {
        QKeyEvent *keyEvent = reinterpret_cast<QKeyEvent*>(event);
        if ((keyEvent->modifiers() & Qt::ControlModifier)
            && (keyEvent->key() == Qt::Key_U)) {
          return true;
        }
      }
      return QDialog::eventFilter(object, event);
    }

the .imp file:

    [
        { symbol: [ "QKeyEvent",                                "private", "<QKeyEvent>", "public" ] },
        { include: [ "@\"(QtCore/)?qcoreevent\\.h\"",           "private", "<QEvent>", "public" ] },
        { include: [ "@\"(QtCore/)?qflags\\.h\"",               "private", "<QFlags>", "public" ] },
        { include: [ "@\"(QtCore/)?qnamespace\\.h\"",           "private", "<Qt>", "public" ] },
        { include: [ "@\"(QtWidgets/)?qdialog\\.h\"",           "private", "<QDialog>", "public" ] },
    ]

batch file

    C:\Apps\include-what-you-use\bin\include-what-you-use.exe -Xiwyu --verbose=7 -Xiwyu --transitive_includes_only -D_MT -D_DLL -m32 -Wno-inconsistent-missing-override --system-header-prefix=Q  -fdiagnostics-format=msvc -fno-show-column -fmsc-version=1800 -D_DEBUG -Xiwyu --mapping_file=test.imp -DUNICODE -DWIN32 -DNOMINMAX -DQT_CORE_LIB -DQT_GUI_LIB -DQT_WIDGETS_LIB -D_UNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -DNDEBUG -DQT_MESSAGELOGCONTEXT -IC:\Apps\Qt\5.4\msvc2013_opengl\mkspecs\win32-msvc2013 -IC:\Apps\Qt\5.4\msvc2013_opengl\include -IC:\Apps\Qt\5.4\msvc2013_opengl\include\QTCore -IC:\Apps\Qt\5.4\msvc2013_opengl\include\QTGui -IC:\Apps\Qt\5.4\msvc2013_opengl\include\QTWidgets test.cpp

**Comment by ThosRTanner** [comment]
This is the (normal) output:


    test.cpp should add these lines:
    #include <QEvent>     // for QEvent, QEvent::Type::KeyPress
    #include <QFlags>     // for QFlags
    #include <Qt>         // for Key::Key_U, KeyboardModifier::ControlModifier
    class QKeyEvent;
    class QObject;

    test.cpp should remove these lines:

    The full include-list for test.cpp:
    #include <QDialog>    // for QDialog
    #include <QEvent>     // for QEvent, QEvent::Type::KeyPress
    #include <QFlags>     // for QFlags
    #include <QKeyEvent>  // for QKeyEvent
    #include <Qt>         // for Key::Key_U, KeyboardModifier::ControlModifier
    class QKeyEvent;
    class QObject;
    ---

**Comment by kimgr** [comment]
Sorry it took me a while to get back to this.

So is the problem that it suggests *both* including and forward-declaring QKeyEvent? Or is something else afoot?

**Comment by ThosRTanner** [comment]
yes, that's what I think is wrong. I think the forward declaration is spurious.

**Comment by jcelerier** [comment]
Can confirm, iwyu adds a lot of useless forward declarations for Qt classes it also actually includes.

For instance: 

    The full include-list for /home/jcelerier/i-score/base/lib/core/view/View.hpp:
    #include <iscore_lib_base_export.h>  // for ISCORE_LIB_BASE_EXPORT
    #include <qobjectdefs.h>             // for Q_OBJECT, signals, slots
    #include <QMainWindow>               // for QMainWindow
    #include <QPair>                     // for QPair
    #include <QString>                   // for QString
    #include <vector>                    // for vector
    class QCloseEvent;  // lines 8-8
    class QDockWidget;  // lines 9-9
    class QEvent;  // lines 10-10
    class QObject;  // lines 11-11
    class QString;
    class QTabWidget;  // lines 12-12
    
See for instance `#include <QString>` followed by `class QString;`

**Comment by bolshakov-a** [comment]
The bug is that IWYU doesn't consider headers from symbol mappings when trying to determine if any of the public headers for a definition is already included or should be included in the file with a fwd-decl use ([see `CalculateIwyuForForwardDeclareUse`](https://github.com/include-what-you-use/include-what-you-use/blob/2ffa17479bbbc183c0abd4bd6fb39a06db58afbd/iwyu_output.cc#L1580)). A simple reproducer without Qt is:
```cpp
// private.h
class Class {};

// public.h
#include "private.h"

// user.cpp
#include "public.h"

Class c, *p;
```
and the mapping file `map.imp`:
```yaml
[
  { symbol: [ "Class", "private", "\"public.h\"", "public" ] },
# { "include": [ "\"private.h\"", "private", "\"public.h\"", "public"] },
]
```
IWYU result is:
```
$ include-what-you-use -Xiwyu --mapping_file=map.imp -I . user.cpp

user.cpp should add these lines:
class Class;

user.cpp should remove these lines:

The full include-list for user.cpp:
#include "public.h"  // for Class
class Class;
```
If you uncomment include-mapping entry in `map.imp`, the bug disappears.

In fact, `qt5_11.imp` mapping file currently delivered with IWYU already contains the needed include-mapping entries, so I cannot reproduce such erroneous behavior with it.
